core/src/api/api.rs

```diff
@@ -135,107 +134,111 @@ impl GraphAPI for GraphState {
 	/// Import raw data retrieved from the blockchain into a user graph.
 	/// Will overwrite any existing graph data for the user,
 	/// but pending updates will be preserved.
-	fn import_user_data(
-		&mut self,
-		ImportBundle { schema_id, pages, dsnp_keys, dsnp_user_id, key_pairs }: ImportBundle,
-	) -> Result<()> {
-		let dsnp_config = self
-			.get_dsnp_config(schema_id)
-			.ok_or(Error::msg("Invalid schema id for environment!"))?;
-		let config = self.environment.get_config();
-		let connection_type = config
-			.get_connection_type_from_schema_id(schema_id)
-			.ok_or(Error::msg("Invalid schema id for environment!"))?;
-		self.public_key_manager
-			.deref()
-			.borrow_mut()
-			.deref_mut()
-			.import_dsnp_keys(dsnp_keys)?;
-
-		let user_graph = match self.user_map.get_mut(&dsnp_user_id) {
-			Some(graph) => graph,
-			None => self.add_user_graph(&dsnp_user_id)?,
-		};
+	// TODO: should make it transactional
+	fn import_users_data(&mut self, payloads: Vec<ImportBundle>) -> Result<()> {
+		for ImportBundle { schema_id, pages, dsnp_keys, dsnp_user_id, key_pairs } in payloads {
+			let dsnp_config = self
+				.get_dsnp_config(schema_id)
+				.ok_or(Error::msg("Invalid schema id for environment!"))?;
+			let config = self.environment.get_config();
+			let connection_type = config
+				.get_connection_type_from_schema_id(schema_id)
+				.ok_or(Error::msg("Invalid schema id for environment!"))?;
+			self.public_key_manager
+				.deref()
+				.borrow_mut()
+				.deref_mut()
+				.import_dsnp_keys(&dsnp_keys)?;
 
-		let user_key_manager = user_graph.user_key_manager_mut();
-		let include_secret_keys = !key_pairs.is_empty();
+			let user_graph = match self.user_map.get_mut(&dsnp_user_id) {
+				Some(graph) => graph,
+				None => self.add_user_graph(&dsnp_user_id)?,
+			};
 
-		// import key-pairs inside user key manager
-		user_key_manager.import_key_pairs(key_pairs);
+			let user_key_manager = user_graph.user_key_manager_mut();
+			let include_secret_keys = !key_pairs.is_empty();
 
-		let resolved_keys = match include_secret_keys {
-			true => user_key_manager.get_all_resolved_keys(),
-			false => vec![],
-		};
+			// import key-pairs inside user key manager
+			user_key_manager.import_key_pairs(key_pairs)?;
 
-		let graph = user_graph.graph_mut(&schema_id);
-		graph.clear();
+			let resolved_keys = match include_secret_keys {
+				true => user_key_manager.get_all_resolved_keys(),
+				false => vec![],
+			};
 
-		match (connection_type.privacy_type(), include_secret_keys) {
-			(PrivacyType::Public, _) => graph.import_public(connection_type, pages),
-			(PrivacyType::Private, true) => {
-				graph.import_private(&dsnp_config, connection_type, &pages, resolved_keys)?;
-				self.pri_manager
+			let graph = user_graph.graph_mut(&schema_id);
+			graph.clear();
```

Comment:
Regarding the TODO above to make this transactional, I suppose one step toward that would be to create a new graph and only replace the existing one when the import completes successfully.

Reply:
We can do that or I was thinking about having a dryrun flag described in https://github.com/LibertyDSNP/graph-sdk/issues/40. Basically we run all the input using **dryrun = true** and if everything looked good we would run it again.

Reply:
Open to both ideas. We can do a quick evaluation of how much is the effort with each one and progress from there. I'll update that ticket to have your approach as an alternative to explore.